File: srcCpp/RTIDDSLoggerDevice.cxx
/*
 * (c) 2005-2018  Copyright, Real-Time Innovations, Inc. All rights reserved.
 * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
 */

#include "RTIDDSLoggerDevice.h"

RTIDDSLoggerDevice::RTIDDSLoggerDevice(): 
        _shmemErrors(false), 
        _zerocopyErrors(false),
        _outOfLoanedSamples(false) {
}

void RTIDDSLoggerDevice::write(const NDDS_Config_LogMessage *message)
{
    if (message) {
        if (message->level == NDDS_CONFIG_LOG_LEVEL_ERROR) {
            if (std::string(message->text).find(
                    NDDS_TRANSPORT_LOG_SHMEM_FAILED_TO_INIT_RESOURCE)
                    != std::string::npos) {
                if (!_shmemErrors) {
                    printf("%s", message->text);
                }
                _shmemErrors = true;
            } else if (std::string(message->text).find(
                    NDDS_TRANSPORT_LOG_SHMEM_SHARED_MEMORY_INSUFFICIENT)
                    != std::string::npos) {

                if (!_shmemErrors) {
                    printf("%s", message->text);
                }

                _shmemErrors = true;
            } else if (std::string(message->text).find(
                    NDDS_ZEROCOPY_NULL_OFFSET_ERROR) != std::string::npos) {

                if (!_zerocopyErrors) {
                    printf("Warning: The Publisher is reusing memory to send different samples before the original samples are processed by the subscriber, leading to inconsistent samples.\n");
                    printf("This is expected with the current perftest configuration. See more on the User manual: 22.5.1.3 Checking data consistency with Zero Copy transfer over shared memory on page 870 .  \n");
                    printf("Unconsistent samples will be reported as lost.\n");
                }

                _zerocopyErrors = true;
            } else if (
                std::string(message->text).find(
                        NDDS_OUT_OF_LOANED_SAMPLES) != std::string::npos ||
                std::string(message->text).find(
                        NDDS_OUT_OF_LOANED_SAMPLES_2) != std::string::npos) {

                if (!_outOfLoanedSamples) {
                    printf("%sThis is expected. Try to reduce it or avoid it by increasing the send queue size.\n", message->text);
                }

                _outOfLoanedSamples = true;
            } else {
                printf("%s", message->text);
            }
        } else {
            printf("%s", message->text);
        }
    }
}

bool RTIDDSLoggerDevice::checkShmemErrors()
{
    return _shmemErrors;
}

void RTIDDSLoggerDevice::close()
{

}

File: srcJava/com/rti/perftest/ddsimpl/RTIDDSLoggerDevice.java
/*
 * (c) 2005-2018  Copyright, Real-Time Innovations, Inc. All rights reserved.
 * Subject to Eclipse Public License v1.0; see LICENSE.md for details.
 */

package com.rti.perftest.ddsimpl;

import com.rti.ndds.config.LoggerDevice;
import com.rti.ndds.config.LogMessage;
import com.rti.ndds.config.LogLevel;

/*
 *  After setting "UPDv4 | SHMEM" as default transport_builtin mask. The user may
 *  get the following error running a default scenario:
 *
 *      [D0001|ENABLE]NDDS_Transport_Shmem_create_recvresource_rrEA:failed to initialize shared memory resource segment for key 0x40894a
 *      [D0001|ENABLE]NDDS_Transport_Shmem_create_recvresource_rrEA:failed to initialize shared memory resource segment for key 0x40894c
 *      [D0001|ENABLE]DDS_DomainParticipantPresentation_reserve_participant_index_entryports:!enable reserve participant index
 *      [D0001|ENABLE]DDS_DomainParticipant_reserve_participant_index_entryports:Unusable shared memory transport. For a more in-depth explanation of the possible problem and solution, please visit http://community.rti.com/kb/osx510.
 *      [D0001|ENABLE]DDS_DomainParticipant_enableI:Automatic participant index failed to initialize. PLEASE VERIFY CONSISTENT TRANSPORT / DISCOVERY CONFIGURATION.
 *      [NOTE: If the participant is running on a machine where the network interfaces can change, you should manually set wire protocol's participant id]
 *      DDSDomainParticipant_impl::createI:ERROR: Failed to auto-enable entity.
 *
 *  By using our own implementation of LoggerDevice, we can filter those errors.
 *  In the case that those errors appear, the execution will be stopped and
 *  a message will be printed showing:
 *      - A link (http://community.rti.com/kb/osx510) about how to solve the issue
 *      - How to setup a different transport via command-line parameters.
 */

 public class RTIDDSLoggerDevice implements LoggerDevice {

    /*
     *   shmemErrors: 'False' by default. In the case that SHMEM issues appear,
     *       it will be set to 'True'.
     */
    private boolean shmemErrors = false;
    private static String NDDS_TRANSPORT_LOG_SHMEM_FAILED_TO_INIT_RESOURCE =
            "NDDS_Transport_Shmem_create_recvresource_rrEA:failed to initialize shared memory resource segment for key";

    /*
     *   @brief This function is the constructor of our internal logging device.
     */
    public RTIDDSLoggerDevice()
    {
    	this.shmemErrors = false;
    }

    /*
     *   @brief This function is used to filter the log messages and write them
     *       through the logger device.
     *       shmemErrors will be set to 'True' if the log message is the known SHMEM issue.
     *   @param message \b In. Message to log.
     */
    public void write(LogMessage message)
    {
        if (!shmemErrors) {
            if (message.level == LogLevel.NDDS_CONFIG_LOG_LEVEL_ERROR) {
                if (message.text.contains(
                        NDDS_TRANSPORT_LOG_SHMEM_FAILED_TO_INIT_RESOURCE)) {
                    shmemErrors = true;
                }
            }
            if (!shmemErrors) {
                System.out.print(message.text);
            }
        }
    }

    /*
     *   @brief Close the logging device.
     */
    public void close()
    {

    }

    /*
     *   @brief Get the value of the variable shmemErrors.
     *   @return shmemErrors
     */
    public boolean checkShmemErrors()
    {
       return this.shmemErrors;
    }
}

// ===========================================================================
